#! /bin/sh -e

# DP: g++ visibility patches (taken from the redhat/gcc-4_1-branch)
# DP:   PR c++/28407, PR c++/28409

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


------------------------------------------------------------------------
r116507 | jakub | 2006-08-28 14:31:26 +0200 (Mon, 28 Aug 2006) | 24 lines

2006-07-20  Jason Merrill  <jason@redhat.com>

        * tree.c (remove_attribute): New fn.

        PR c++/28407
        * decl.c (grokvardecl): Set DECL_THIS_STATIC on file-scope
        const variables with implicit internal linkage.
        * tree.c (decl_linkage): Only return lk_external if it's set.

        PR c++/28409
        * decl2.c (constrain_visibility): Ignore the anonymous namespace 
        for extern "C" decls.
        (VISIBILITY_STATIC): Rename to VISIBILITY_ANON.

        * decl2.c (constrain_visibility): Remove specified and reason 
        parameters.  Don't touch decls that already have explicit visibility.
        (determine_visibility): Do copy DECL_VISIBILITY_SPECIFIED from 
        template.
        (determine_visibility_from_class): Reverse sense of 
        DECL_VISIBILITY_SPECIFIED test for target-specific visibility rules.
        (constrain_class_visibility): Only complain about member visibility
        if the member type is another class.  Don't change visibility of the
        current class.

Index: gcc/doc/extend.texi
===================================================================
--- gcc/doc/extend.texi	(Revision 116506)
+++ gcc/doc/extend.texi	(Revision 116507)
@@ -2346,6 +2346,9 @@
 Otherwise, template instantiations and specializations default to the
 visibility of their template.
 
+If both the template and enclosing class have explicit visibility, the
+visibility from the template is used.
+
 @item warn_unused_result
 @cindex @code{warn_unused_result} attribute
 The @code{warn_unused_result} attribute causes a warning to be emitted
@@ -3465,6 +3468,13 @@
 attributes, the attribute must appear between the initial keyword and
 the name of the type; it cannot appear after the body of the type.
 
+Note that the type visibility is applied to vague linkage entities
+associated with the class (vtable, typeinfo node, etc.).  In
+particular, if a class is thrown as an exception in one shared object
+and caught in another, the class must have default visibility.
+Otherwise the two shared objects will be unable to use the same
+typeinfo node and exception handling will break.
+
 @subsection ARM Type Attributes
 
 On those ARM targets that support @code{dllimport} (such as Symbian
